### PR TITLE
#310: Only properties created by maven-git-commit-id-plugin filtered.

### PR DESCRIPTION
**Pull request by cbuschka**
Added extra checks if a property has been generated by maven-git-commit-id-plugin before filtering. This fixes issue 310.

### CODE REVIEW

**Comment by TheSnoozer** [comment]
Hi thank you for reporting the issue and creating a pull request!
This is indeed kinda ugly....I'm just thinking if there would be other things effected when using the outlined scenario...for example in the latest version of the plugin has an option to replace properties...however using the outlined scenario the replacement of properties would work differently and thus the question raises why we even use different property sets inside initProperties() see
https://github.com/ktoso/maven-git-commit-id-plugin/blame/master/src/main/java/pl/project13/maven/git/GitCommitIdMojo.java#L538.

The PR looks ok, but I'll definitely need to look into that further to check if other items of the plugin would be affected by this as well. Also would be nice to have a testcase for this since it is something more critical ;-)

**Comment by cbuschka** [comment]
Yes you are right I will create a unit test for.

**Comment by cbuschka** [comment]
Ok, I moved the filtering into a collaborator (PropertiesFilterer) and added a unit test.

**Comment by TheSnoozer** [comment]
Thanks for the test cases!
Yes, fully agree! The `properties = initProperties();` should rather be always `properties = new Properties()`.

It seems that the `project.getProperties().putAll(properties);` here https://github.com/ktoso/maven-git-commit-id-plugin/blob/master/src/main/java/pl/project13/maven/git/GitCommitIdMojo.java#L383 is independent from the `generateGitPropertiesFile` switch.
Currently the initProperties get the mavenProject-properties when `generateGitPropertiesFile=false`
and if `generateGitPropertiesFile=true` the properties are being attached with the `project.getProperties().putAll(properties);` outlined above.

Please add the refactoring and I'll pull in the changes....
Thanks again for reporting and creating a PR :-)

**Comment by cbuschka** [comment]
Moved replacement code into PropertiesReplacer with own test and added the refactoring described above. The refactoring changes behaviour! What is the intend behind the properties replacement? Should it be performed on project wide properties?

**Comment by TheSnoozer** [comment]
Actually its a really good question on what properties the replacement should perform. The initial intent comes from here https://github.com/ktoso/maven-git-commit-id-plugin/issues/138. In a nutshell a branch has a name like "feature/feature_name" and someone wanted to use the branch-name in the finalName of the artifact. However the character "/" is not valid as filename and thus this would not work.

Based on what I wrote inside the docs (https://github.com/ktoso/maven-git-commit-id-plugin/commit/1e8b523cddf0d44ca664e1e158fa80f0866acb4f) the replacement can be used to replace certain characters or strings using regular expressions within the generated git properties. I even added the following note `Please note that the replacement will *only be applied to properties that are being generated by the plugin.`...

I would say at the time being it is sufficient to have the replacement only work on the properties generated by the plugin (that is pretty much what the readme says about the functionality).

Just for reference:
We either can adjust the way it works (it requested with a string reason) or the user can use 
https://code.google.com/archive/p/maven-replacer-plugin/ which is pretty much the same but for all properties.

TLDR
======
All changes look good to me!
Thanks for the code cleanup!
I will pull in the changes since the replacement should currently only work on properties generated by the plugin (was not released yet, so no one can claim it ever worked differently :-P)

**Comment by cbuschka** [comment]
Thanks for merging! I am glad I could contribute and give something back in return.

**Comment by ktoso** [comment]
Thanks @cbuschka !

This is now released as 2.2.3, have fun! :)
